docs/presentation/slides.md

```diff
@@ -0,0 +1,507 @@
+  let clocks = rcu.cfgr.sysclk(72.mhz()).adcclk(12.mhz()).freeze(&mut flash.ws);
+
+  let mut watchdog = FreeWatchdog::new(dp.FWDGT);
+  watchdog.start(WATCHDOG_MILLIS.ms());
+
+  let systick = SysTick::start(cp.SYST, &clocks);
+
+  let mut hoverboard = Hoverboard::new(dp.GPIOA, dp.TIMER0, &mut rcu.apb1,
+                                       clocks, ...);
+
+  // More initialisation...
+
+  loop {
+    watchdog.feed();
+
+    // Process events...
+  }
+}
+```
+
+???
+
+- `Peripherals::take()` returns a struct containing the PAC representations of all peripherals. If
+  you call it a second time it returns `None`, to prevent multiple references to the same registers
+  existing.
+- `constrain()` comes from extension traits which essentially convert the PAC type into the HAL
+  type.
+- `SysTick` gives us a clock, so we can measure time durations.
+
+---
+
+# Abstraction
+
+Let's make a wrapper for the peripherals on the hoverboard.
+
+```rust
+pub struct Hoverboard {
+  pub serial_rx: Rx<USART1>,
+  pub serial_writer: BufferedSerialWriter<Tx<USART1>>,
+  pub imu: Bmi160<I2cInterface<
+      BlockingI2c<I2C0, PB8<Alternate<AF1>>, PB9<Alternate<AF1>>>>>,
+  pub buzzer: Buzzer,
+  pub power_latch: PB2<Output<PushPull>>,
+  /// This will be high when the power button is pressed.
+  pub power_button: PC15<Input<Floating>>,
+  pub leds: Leds,
+  // More fields...
+}
+```
+
+???
+
+- `Bmi160` comes from another library, which gives us access to the IMU over I2C.
+- `Leds` is a struct with appropriately named fields for the various LEDs connected to GPIO pins.
+- `Buzzer` is a small wrapper around the PWM module from the HAL, which uses a timer to drive the
+  piezo speaker at a given frequency.
+- There are more fields for the other USART, charge state, and some configuration.
+- There is also a struct representing the motor, but it is stored in a static variable inside a
+  `Mutex` because it needs to be accessed from interrupt handlers.
+
+---
+
+# BLDC motor driver
+
+![BLDC motor driver circuit](bldc_driver.svg)
+
+???
+
+- Three motor windings connected between pairs of MOSFETs, like an H-bridge with an extra leg.
+- Each leg can be driven high, driven low, or undriven.
+
+---
+
+# BLDC motor control
+
+| Position | Sensor values | Forwards | Backwards |
+| -------- | ------------- | -------- | --------- |
+| 0        | `0 0 1`       | `0 + -`  | `0 - +`   |
+| 1        | `1 0 1`       | `- + 0`  | `+ - 0`   |
+| 2        | `1 0 0`       | `- 0 +`  | `+ 0 -`   |
+| 3        | `1 1 0`       | `0 - +`  | `0 + -`   |
+| 4        | `0 1 0`       | `+ - 0`  | `- + 0`   |
+| 5        | `0 1 1`       | `+ 0 -`  | `- 0 +`   |
+
+???
+
+- We have three Hall effect sensors connected to digital inputs, and three motor windings.
+- We drive all the motor connections from a single PWM unit, configured to use a different duty
+  cycle for each pair of outputs.
+- '0' here actually means a 50% duty cycle, '+' means > 50%, '-' means < 50%.
```

Comment:
I did not appreciate this. I should look over the code + circuit diagrams again.